Fix and test constant scalar Records

anokfireball opened the pull request:
For constant scalar `Record`s, both the `Record` and the contained sclar `RecordComponent` are the same group in the persistent datastructure. The `RecordComponent`has not been properly assigned for `Mesh`es, which made it impossible to read back files with constant scalar `MeshRecordComponent`s.

This addresses the issue by assigning the group to both the `Record` and the `RecordComponent` for both `Mesh` and `ParticleSpecies`.

Fix #356.

anokfireball committed:
Fix and test constant scalar Records

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -1445,6 +1445,40 @@ TEST_CASE( "hdf5_110_optional_paths", "[serial][hdf5]" )
     }
 }
+
+TEST_CASE( "hdf5_constant_scalar", "[serial][hdf5]" )
```
@C0nsultant should we also add a test for a non-scalar record that only has a constant record component (but not all record components are constant)? Or is this case already covered somewhere else?

anokfireball replied:
Are you talking about a `Record`with mixed constant and non-constant `RecordComponent`s? I don't think we have such a case yet.

ax3l replied:
Yep, that's what I refer to.

anokfireball replied:
I included a little something for this.

ax3l commented on src/ParticleSpecies.cpp:
```diff
@@ -58,18 +58,17 @@ ParticleSpecies::read()
-            auto value = std::find(attrBegin, attrEnd, "value");
-            auto shape = std::find(attrBegin, attrEnd, "shape");
-            if( value != attrEnd && shape != attrEnd )
+            auto begin = aList.attributes->begin();
```
```
error: declaration of ‘begin’ shadows a member of 'this' [-Werror=shadow]
error: declaration of ‘end’ shadows a member of 'this' [-Werror=shadow]
```

ax3l reviewed (commented):
Annotated the shadowing Werror inline.
Feel free to also update the changelog with the fix for it.

anokfireball committed:
Test case for mixed constant/non-constant

anokfireball committed:
Changelog

ax3l commented:
@C0nsultant sorry, changelog rebase :)

ax3l committed:
SerialIOTest: +ADIOS1 Constant Scalar

Generalize the test for constant record components for both "bp" and
"h5" backend.

ax3l committed:
ADIOS1: Attribute string array handling

Add support for string array handling in ADIOS1 attribute
reading.

ax3l commented:
@C0nsultant I extended the test to also test the "bp" ADIOS1 backend. First I noted: attribute reading for string arrays (such as `axisLabels`) was missing, added it.

What is now missing: in ADIOS1, all attributes, records, etc. are just addressed by a simple map with the path as key. For constant, *scalar* record components, there seem to be the "directory" of the scalar record missing. Afair, ADIOS1 does not have any directories (see above) anyway, so we might need to add extra handling for this to emulate /  properly detect it anyway.

Concrete, there are attributes such as
```
  double              /data/0/meshes/char/gridGlobalOffset        attr
  unsigned long long  /data/0/meshes/char/shape                   attr
  string              /data/0/meshes/char/geometry                attr
  double              /data/0/meshes/char/gridSpacing             attr
  double              /data/0/meshes/char/unitSI                  attr
  byte                /data/0/meshes/char/value
// ...
```

in a scalar constant record, but the empty `/data/0/meshes/char/` is missing which we query for in `Iteration.cpp` and `ParticleSpecies.cpp`.

This can probably be fixed in `CommonADIOS1IOHandlerImpl::listPaths`.

ax3l commented on CHANGELOG.rst:
```diff
@@ -26,13 +26,14 @@ Bug Fixes
 """""""""
 
 - dataOrder: mesh attribute is a string #355
+- constant scalar MeshRecords: reading corrected #358
```
note: also constant scalar ParticleRecords. so maybe `constant scalar records: reading corrected #358` ?

anokfireball replied:
The logic for `ParticleRecords` was actually correct, it just did manually what the second `READ_PATH` now does automatically. It mainly is there to simplify and unify the code in both.

anokfireball committed:
Fix ADIOS1 constant scalar detection

As empty groups can not be stored in ADIOS, we have to work around this
by detecting whether an attribute is at a location that is NOT a
variable.

anokfireball commented:
> This can probably be fixed in `CommonADIOS1IOHandlerImpl::listPaths`

That's the only place where it can be fixed without major changes. I hacked together a solution and while I don't think it's too elegant... there's probably not many other ways to do this. Essentially, we have to verify that something is a group by checking that it is not a variable :fearful:.

ax3l commented:
Thank you, that's great! IMHO, for the ADIOS1 implementation fully lacking hierarchy awareness, your detection ist very gentle now. Love it!
